frostsnapp/lib/restoration.dart

```diff
@@ -1918,26 +2023,109 @@ class _PhysicalBackupFailView extends StatelessWidget with _TitledWidget {
   Widget build(BuildContext context) {
     final String compatMessage = switch (compatibility) {
       ShareCompatibility_ConflictsWith(:final deviceId, :final index) =>
-        "You have already restored backup #$index on ‘${coord.getDeviceName(id: deviceId)!}’ and it doesn't match the one you just entered. Are you sure that this backup is for this wallet?",
-      _ => "Unknown error of kind: $compatibility",
+        "You have already restored backup #$index on '${coord.getDeviceName(id: deviceId)!}' and it doesn't match the one you just entered. Try a different backup.",
+      ShareCompatibility_Incompatible() =>
+        "This backup is for a different wallet. Try a different backup.",
+      ShareCompatibility_AlreadyGotIt() =>
+        "You've already added this backup to the restoration. Try a different backup.",
+      _ => "The backup is not compatible with this wallet.",
     };
 
     final String message = errorMessage ?? compatMessage;
     return MaterialDialogCard(
       key: const ValueKey('physicalBackupFail'),
       iconData: Icons.error_rounded,
-      title: Text('Couldn\'t load backup'),
+      title: Text('Incompatible backup'),
       content: Text(message, textAlign: TextAlign.center),
-
       actions: [
         TextButton(child: const Text('Close'), onPressed: onClose),
-        FilledButton(child: const Text('Retry'), onPressed: onRetry),
+        FilledButton(
+          child: const Text('Try Different Backup'),
+          onPressed: onRetry,
+        ),
       ],
     );
   }
 
   @override
-  String get titleText => '';
+  String get titleText => 'Backup Error';
+}
+
+class _FirmwareUpgradeView extends StatefulWidget with _TitledWidget {
+  final TargetDevice targetDevice;
+  final VoidCallback onComplete;
+  final VoidCallback onCancel;
+
+  const _FirmwareUpgradeView({
+    super.key,
+    required this.targetDevice,
+    required this.onComplete,
+    required this.onCancel,
+  });
+
+  @override
+  State<_FirmwareUpgradeView> createState() => _FirmwareUpgradeViewState();
+
+  @override
+  String get titleText => 'Firmware Upgrade Required';
+}
+
+class _FirmwareUpgradeViewState extends State<_FirmwareUpgradeView> {
+  late final DeviceActionUpgradeController _controller;
+  bool _isUpgrading = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _controller = DeviceActionUpgradeController();
+    // Don't handle disconnection - firmware upgrade causes device to reset which is normal
```

Comment:
@musdom found an issue where the phone-connected device shows a black screen on android after a firmware update. Not sure if related.